adapters/relevantdigital/relevantdigital.go

```diff
@@ -0,0 +1,278 @@
+	count, cerr := jsonparser.GetInt(prebidBidRequest.Ext, "relevant", "count")
+	if cerr != nil {
+		count = 0
+	}
+
+	if count >= 5 {
+		return &errortypes.FailedToRequestBids{
+			Message: "too many requests",
+		}
+	} else {
+		count = count + 1
+	}
+
+	debug, derr := jsonparser.GetBoolean(prebidBidRequest.Ext, "prebid", "debug")
+	if derr != nil {
+		debug = false
+	}
+
+	prebidBidRequest.Ext = []byte(fmt.Sprintf(STORED_REQUEST_EXT, debug, id, count))
+	return nil
+}
+
+func patchBidImpExt(imp *openrtb2.Imp, id string) {
+	imp.Ext = []byte(fmt.Sprintf(STORED_IMP_EXT, id))
+	if imp.Banner != nil {
+		imp.Banner.Ext = nil
+	}
+	if imp.Video != nil {
+		imp.Video.Ext = nil
+	}
+	if imp.Native != nil {
+		imp.Native.Ext = nil
+	}
+	if imp.Audio != nil {
+		imp.Audio.Ext = nil
+	}
+}
+
+func setTMax(prebidBidRequest *openrtb2.BidRequest, pbsBufferMs int) {
+	timeout := float64(prebidBidRequest.TMax)
+	if timeout <= 0 {
+		timeout = DEFAULT_TIMEOUT
+	}
+	buffer := float64(pbsBufferMs)
+	prebidBidRequest.TMax = int64(math.Min(math.Max(timeout-buffer, buffer), timeout))
+}
+
+func cloneBidRequest(prebidBidRequest *openrtb2.BidRequest) (*openrtb2.BidRequest, error) {
+	jsonRes, err := json.Marshal(prebidBidRequest)
+	if err != nil {
+		return nil, err
+	}
+	var copy openrtb2.BidRequest
+	err = json.Unmarshal(jsonRes, &copy)
+	return &copy, err
+}
+
+func createBidRequest(prebidBidRequest *openrtb2.BidRequest, params []*openrtb_ext.ExtRelevantDigital) (*openrtb2.BidRequest, error) {
+	bidRequestCopy, err := cloneBidRequest(prebidBidRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	err = patchBidRequestExt(bidRequestCopy, params[0].AccountId)
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("failed to create bidRequest, error: %s", err),
+		}
+	}
+
+	setTMax(bidRequestCopy, params[0].PbsBufferMs)
+
+	for idx := range bidRequestCopy.Imp {
+		patchBidImpExt(&bidRequestCopy.Imp[idx], params[idx].PlacementId)
+	}
+	return bidRequestCopy, err
+}
+
+func getImpressionExt(imp *openrtb2.Imp) (*openrtb_ext.ExtRelevantDigital, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "imp.ext not provided",
+		}
+	}
+	relevantExt := openrtb_ext.ExtRelevantDigital{PbsBufferMs: DEFAULT_BUFFER_MS}
+	if err := json.Unmarshal(bidderExt.Bidder, &relevantExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}
+	}
+	return &relevantExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtRelevantDigital) (string, error) {
+	params.Host = strings.ReplaceAll(params.Host, "http://", "")
+	params.Host = strings.ReplaceAll(params.Host, "https://", "")
+	params.Host = strings.ReplaceAll(params.Host, RELEVANT_DOMAIN, "")
+
+	endpointParams := macros.EndpointTemplateParams{Host: params.Host}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
```

Comment:
https://github.com/prebid/prebid-server/pull/3216/files#r1366519731

Reply:
Yes, we have discussed this previously in this PR and in the Pbjs PR.

Heres my previous explanation:
```
For consistency between our adapters, I hope we could allow dynamic subdomains just as discussed here: https://github.com/prebid/Prebid.js/pull/9685#discussion_r1159833733

The main problem is that existing customers have prebid server clusters running at certain subdomains. Adding the proposed method in https://github.com/prebid/prebid-server/issues/2606 of query parameters and load balancer endpoint would add an extra request and that will have negative impact on the customers latency.
```

During Prebid Server PMC, I asked about this as well. And the consensus was that, yes, its not ideal (from a hosting perspective). However, it should be ok to have a "semi" dymanic host, where the TLD is fixed.

Reply:
The endpoint domain rules are:
- They cannot be fully dynamic
- If the dynamic part is geographic, a regional 'host' parameter cannot be **required**. It may be optional. The vastly better solution to regional endpoints is that host companies deploy different static endpoints in their various datacenters.
- Host companies are advised to watch for issues related to abuse of dynamic domains and be ready to disable that adapter if needed

Reply:
FWIW, see https://github.com/prebid/prebid-server/pull/3311 as my proposal for how bidders can document their regional endpoints in the YAML for host company utilization.

Reply:
> Thanks @bretg for the clarification. I hope we can mark this as resolved and allow the current PR semi dynamic host.

@hugolm84 

Major concerns with such semi dynamic usage are,

- security concerns
   The security aspect is alleviated by using a fixed top level domain. Due to the potential harm to hosts, we are strict in this requirement. We are working towards fixing the few adapters currently in violation.

- connection performance
  The connection performance advice is for your benefit. Client specific subdomains prevent Prebid Server from reusing connections across your clients which results in more connections needed for your adapter. The issue gets worse the more successful you are in attracting new clients. Hosts may choose to disable your adapter due to this behavior.

Reply:
@onkarvhanumante - I believe relevantDigital uses 'semi-dynamic' hosts to handle geographic regions, not one per publisher. As long as this code doesn't **require** the host param, we're good. PBS Host Companies can "Do The Right Thing", which is deploy static endpoints.

Reply:
Hi @bretg @onkarvhanumante!

I understand your concerns, however, I believe the semi-dynamic host should be allowed in our case as, quoting @bretg, the rules are:

```
The endpoint domain rules are:
   They cannot be fully dynamic
```

We do not use the host param for geographic routing. Each customer have their own CNAME pointing to a prebid-server residing in one or several clusters. Geographical routing is done on DNS level. Load-balancing is done within the cluster.

The reason each customer have their own subdomain, is that, to name a few:

1. Some customer cannot be allowed to share servers with other customers.
2. We have many stored requests, but some stored requests does not exist on all servers.
3. We need to be able to upgrade only a customer a specific prebid-server version, without impacting any other customers.
4. Allow customers to add or remove clusters as they grow/shrink.

On web, we own the script and we can control the domains customers need to access prebid-servers. However, on Apps, AMP or when someone else hosts the adapter, we can not control the subdomain without asking the app owner or host to update the domain. Therefore, a customer is given a domain that they can use and hardcode into their app. And we can resize the clusters that the domain is pointing to or what ever is needed.

I realise that this can be accomplished by using a "global" load balancer that uses an account id as query parameter and do a lookup on that id to decide on what "clusters" a request should be balanced between. However, this adds latency and breaks the requirement of "some customers can not mix".

Doing a quick grep shows that a couple of the adapters use semi dynamic hosts and these are the ones that required the host parameter:

```
./adapters/smartyads/smartyads.go:      endpointParams := macros.EndpointTemplateParams{Host: params.Host, SourceId: params.SourceID, AccountID: params.AccountID}
./adapters/trafficgate/trafficgate.go:          urlParams := macros.EndpointTemplateParams{Host: reqExt.Host}
./adapters/zeroclickfraud/zeroclickfraud.go:            urlParams := macros.EndpointTemplateParams{Host: reqExt.Host, SourceId: strconv.Itoa(reqExt.SourceId)}
./adapters/bidmachine/bidmachine.go:    endpointParams := macros.EndpointTemplateParams{Host: params.Host}
./adapters/between/between.go:  return macros.ResolveMacros(a.EndpointTemplate, macros.EndpointTemplateParams{Host: e.Host, PublisherID: e.PublisherID})
./adapters/relevantdigital/relevantdigital.go:  endpointParams := macros.EndpointTemplateParams{Host: params.Host}
./adapters/acuityads/acuityads.go:      endpointParams := macros.EndpointTemplateParams{Host: params.Host, AccountID: params.AccountID}
```

 These do it via some other required param to build the URL:
 
 ```
 ./adapters/yeahmobi/yeahmobi.go:        return macros.ResolveMacros(a.EndpointTemplate, macros.EndpointTemplateParams{Host: "gw-" + url.QueryEscape(ext.ZoneId) + "-bid.yeahtargeter.com"})
./adapters/tappx/tappx.go:      endpointParams := macros.EndpointTemplateParams{Host: tappxHost}
./adapters/adocean/adocean.go:  endpointParams := macros.EndpointTemplateParams{Host: params.EmitterPrefix}
 ```

IMHO, if these are allowed, I think that our Semi Dynamic host also should be allowed.

Reply:
Thanks for the explanation @hugolm84 .  Though Prebid does not favor client-specific endpoints, and we warn Prebid Server host companies to watch for problems with these adapters, client-specific semi-dynamic endpoints are not rejected.

Reply:
@onkarvhanumante I believe this means we are good, semi dynamic host wise?